PNG: Add EXIF handling

cjee21 opened the pull request:
Just same thing as for JPEG.

JeromeMartinez reviewed (dismissed):
As with e.g. XMP or C2PA, EXIF needs to have its own files so we can expand it, rather than copying the code (which actually currently does nothing, it was a started by me but I never went further).

cjee21 commented:
Yes.

After I see the new discussion issue, I was thinking of editing this PR to split the Exif part out and add it to PNG to lay ground work for Exif parsing.

cjee21 commented:
It looks similar to parsing TIFF? In that case do we share codes between them?

JeromeMartinez commented:
It is the same base, but like with C2PA similar to MP4, for the moment better to copy the style of TIFF in new files, I may develop in the future something for similar formats but not yet well done so better to copy for the moment.

cjee21 commented:
A probably temporary commit, just trying to do some parsing to understand the structure of EXIF.

cjee21 commented:
Implemented some initial parsing by copying things from Tiff. There is an issue where there is gap between some items. Parsing is broken from that point on at the moment.

JeromeMartinez commented:
> There is an issue where there is gap between some items. 

Sample file please.

cjee21 commented:
> > There is an issue where there is gap between some items.
> 
> Sample file please.

[Untitled.zip](https://github.com/user-attachments/files/19905690/Untitled.zip)

cjee21 commented:
I worked-around the gap/padding/junk issue for now. Will let you see how to handle it better and also might need more checks for unexpected cases.

Now managed to parse basically all EXIF data that is present in GIMP-generated file including in EXIF sub IFD (will leave it to you or others to refine handling of that too).

I think if it is okay after you touch-up, can get this PR done/merged and others (bitplane?) might want to continue work on it.

cjee21 commented:
@JeromeMartinez It seem there is issue with seeking causing the bugs.

Now managed to parse basically all EXIF data that is present in GIMP-generated file including in EXIF sub IFD (will leave it to you or others to refine handling of that too). Also parsed some EXIF data from camera photo including GPS sub IFD.

PS Deleted previous outdated comment.

cjee21 commented:
Need a better way to handle the SubIFDs that use same IDs for different tags too.

cjee21 commented:
Trying to fix the seeking. It appears to work now but I am not sure what I did or how it works 😅

cjee21 commented:
Okay, I think I'll stop here and let others refine and continue work on it. After that I can help test and contribute again after the architecture/ground work is confirmed/stable.

JeromeMartinez commented:
> Okay, I think I'll stop here and let others refine and continue work on it. 

Thank you, I am checking today/tomorrow.

JeromeMartinez commented:
@cjee21 I have some tweaks to add to this PR but I can confirm it works pretty well, but I'll need time for validating with more files, for instance I added to my local branch the handling of Exif in HEIC for testing my smartphone and I see the GPS data and model name but alo even more metadata, I try to support all Exif tags I find on more or less standard SF/HW before merging the PR.

cjee21 commented:
I have some smartphone samples here (GPS tagging disabled).
[EXIF_samples.zip](https://github.com/user-attachments/files/19935548/EXIF_samples.zip)

For various cameras, samples can be obtained from the link below so I don't think I need to share samples from my cameras.
https://www.dpreview.com/sample-galleries?category=cameras

JeromeMartinez commented:
> For various cameras, samples can be obtained from the link below so I don't think I need to share samples from my cameras.

Wow, a great list of cameras, very useful!
It will take time to manage well all tags (tools like Exiftool were not made in one day), but at least we can check well the first tags we support, and at long term compare our output with Exiftool

cjee21 commented:
@JeromeMartinez 
I have an idea to improve the parsing so I added a commit. It now parses the sub IFDs in correct order so that MediaTrace address does not jump around. Also better handle the different sub IFDs to prevent conflicting tag IDs.

If it conflicts with what you are working on then just put this commit aside first else can merge with previous parsing commit.

cjee21 commented:
Added tag names for the thumbnail IFD and also name the IFDs for clearer view.

JeromeMartinez commented:
UWP project updated
More health checks
Tags are clearly separated by "namespace" also in the info storage part (there could be a tag conflict between namespaces)
More tags listed, descriptions from specs as much has possible
Unique code for IFD offset integrity check
Added SignedRational (found files with apertures etc) & UserComment
color space: it is more complicated, matrix coeeficients etc are specific tags, I removed it for the moment, I need to think about it.
I squashed some of your commits and also my updates for having not so many commits, please pull/force-push.
Please use this version for updates
Added HEIF related support.

JeromeMartinez commented:
Also: Exif thumbnail has its own Image track.

cjee21 commented:
I added timezone to DateTimeOriginal.

JeromeMartinez commented on Source/MediaInfo/Tag/File_Exif.cpp:
```diff
@@ -0,0 +1,1356 @@
+        if (SubSecTimeOriginal != Infos_Exif.end())
+            datetime += __T(".") + SubSecTimeOriginal->second.Read();
+        if (OffsetTimeOriginal != Infos_Exif.end())
+            datetime += __T(" UTC") + OffsetTimeOriginal->second.Read();
```
No " UTC" in ISO 8601.

JeromeMartinez commented on Source/MediaInfo/Tag/File_Exif.cpp:
```diff
@@ -0,0 +1,1356 @@
+        if (GPSLatitude != Infos_GPS.end() && GPSLatitude->second.size() == 3 && GPSLatitudeRef != Infos_GPS.end() && GPSLongitude != Infos_GPS.end() && GPSLongitude->second.size() == 3 && GPSLongitudeRef != Infos_GPS.end()) {
+            Ztring location = GPSLatitude->second.at(0) + __T("\u00B0") + GPSLatitude->second.at(1) + __T("'") + GPSLatitude->second.at(2) + __T("\"") + GPSLatitudeRef->second.at(0) + __T(" ")
+                + GPSLongitude->second.at(0) + __T("\u00B0") + GPSLongitude->second.at(1) + __T("'") + GPSLongitude->second.at(2) + __T("\"") + GPSLongitudeRef->second.at(0) + __T(" ");
+            Fill(Stream_General, 0, General_Recorded_Location, location);
```
Consider to add altitude (optional, like what you did with TimeOriginal).

JeromeMartinez commented:
> https://www.dpreview.com/sample-galleries?category=cameras

I am interested in all galleries from this website, but it prevents the download of a package or the usage of classic web crawlers, do you have a tip for getting all files (jpg but also raw files when available) for avoiding me to spend much time on a trying with dedicated code?

JeromeMartinez commented:
@cjee21 I rebased, not yet full QA.

cjee21 commented:
Looks okay. It now has size for thumbnail image.

JeromeMartinez commented:
Rebased. Still in QA on my side.

cjee21 commented:
@JeromeMartinez do you still need help regarding https://github.com/MediaArea/MediaInfoLib/pull/2276#issuecomment-2841035336?

JeromeMartinez commented:
> do you still need help regarding https://github.com/MediaArea/MediaInfoLib/pull/2276#issuecomment-2841035336?

If you are motivated to make the script, I am interested :-p.
(I did not check yet, focusing on the file I already have first)

cjee21 commented:
> > do you still need help regarding [#2276 (comment)](https://github.com/MediaArea/MediaInfoLib/pull/2276#issuecomment-2841035336)?
> 
> If you are motivated to make the script, I am interested :-p.

[dpreview batch download.zip](https://github.com/user-attachments/files/20208411/dpreview.batch.download.zip)

Fully AI-generated script so use at own risk :p I just insert the comment at first line.
It works the last time I tested it like 2 weeks ago. You have to save the HTML using a browser first (make sure the page saved contains the table of gallery links) then feed the HTML to it if you want to download every gallery.

JeromeMartinez commented:
Lot of incoherence between several metadata source and the metadata inside Exif itself, I give up curating everything now and prioritize to merge what we already have.

cjee21 commented:
Added a commit to prevent duplicate company name in output. Canon cameras have Canon in their model string already. Also fixed the GCC error in CI (https://github.com/MediaArea/MediaInfoLib/actions/runs/15242036391) along with two MSVC warnings (using uninitialized var and unchecked lower bound).

cjee21 commented:
> while keeping the original information as the precision is linked to the way to represent the information

Might want to have another look at the `loci` handling that I wrote too. Looks like in some cases the value is slightly different from other apps. Not sure which is more precise.

JeromeMartinez commented:
> Looks like in some cases the value is slightly different from other apps.

Example?

JeromeMartinez commented:
> Added a commit to prevent duplicate company name in output. Canon cameras have Canon in their model string already

I agree on the goal (it is part of the curation of Exif metadata, and there will be a lot of things like that to do...), but with your method only the displayed string is modified, not `Encoded_Hardware_Model`, so I change the code for modifying also `Encoded_Hardware_Model`.
